massa-consensus-worker/src/controller.rs

```diff
@@ -87,41 +91,62 @@ impl ConsensusController for ConsensusControllerImpl {
     /// Get a part of the graph to send to a node so that he can setup his graph.
     /// Used for bootstrap.
     ///
+    /// # Arguments:
+    /// * `cursor`: streaming cursor containing the current state of bootstrap and what blocks have been to the client already
+    /// * `execution_cursor`: streaming cursor of the final state to ensure that last slot of the bootstrap info corresponds
+    ///
     /// # Returns:
     /// A portion of the graph
-    fn get_bootstrap_graph(&self) -> Result<BootstrapableGraph, ConsensusError> {
+    fn get_bootstrap_part(
+        &self,
+        mut cursor: StreamingStep<PreHashSet<BlockId>>,
+        execution_cursor: StreamingStep<Slot>,
+    ) -> Result<(BootstrapableGraph, StreamingStep<PreHashSet<BlockId>>), ConsensusError> {
+        let mut final_blocks: Vec<ExportActiveBlock> = Vec::new();
+        let mut retrieved_ids: PreHashSet<BlockId> = PreHashSet::default();
         let read_shared_state = self.shared_state.read();
-        let mut required_final_blocks: PreHashSet<_> =
+        let required_blocks: PreHashSet<BlockId> =
             read_shared_state.list_required_active_blocks()?;
-        required_final_blocks.retain(|b_id| {
-            if let Some(BlockStatus::Active { a_block, .. }) =
+
+        let (difference, previous_ids) = match cursor {
+            StreamingStep::Started => (required_blocks, PreHashSet::default()),
+            StreamingStep::Ongoing(ref cursor_ids) => (
+                required_blocks.difference(cursor_ids).cloned().collect(),
+                cursor_ids.clone(),
+            ),
+            StreamingStep::Finished(_) => return Ok((BootstrapableGraph { final_blocks }, cursor)),
+        };
+
+        for b_id in &difference {
+            if let Some(BlockStatus::Active { a_block, storage }) =
                 read_shared_state.block_statuses.get(b_id)
             {
+                if final_blocks.len() as u64 >= self.bootstrap_part_size {
+                    break;
+                }
+                if let StreamingStep::Finished(Some(slot)) = execution_cursor {
+                    if a_block.slot > slot {
```

Comment:
Why this check is only if execution cursor is finished ?

Reply:
Because we want no blocks after the execution cursor once the final state bootstrap is done

Reply:
Ah do you mean that we could potentially get blocks that are in the future of the upcoming last execution slot here while execution streaming is ongoing? If that is what you meant you have a point that is not what we want